x/evm/keeper/migrate.go

```diff
@@ -3,68 +3,115 @@ package keeper
-				return err
-			}
-
-			k.Logger(ctx).Info(fmt.Sprintf("migrated all burner info keys for chain %s", chain.Name))
+		for _, chain := range slices.Filter(n.GetChains(ctx), types.IsEVMChain) {
+			ck := funcs.Must(bk.ForChain(ctx, chain.Name)).(chainKeeper)
 
+			migrateDeposits(ctx, ck, types.DepositStatus_Confirmed)
+			migrateDeposits(ctx, ck, types.DepositStatus_Burned)
 		}
-		k.Logger(ctx).Info("burner info keys migration complete")
+
 		return nil
 	}
 }
 
-func migrateBurnerInfoForChain(ctx sdk.Context, k *BaseKeeper, chain exported.Chain, oldKey string) error {
-	ck, err := k.forChain(ctx, chain.Name)
-	if err != nil {
-		return err
+func migrateDeposits(ctx sdk.Context, ck chainKeeper, status types.DepositStatus) {
+	var iteratedDepositCount, ignoredDepositCount uint64
+
+	var prefix key.Key
+	switch status {
+	case types.DepositStatus_Confirmed:
+		prefix = key.FromStr(confirmedDepositPrefixDeprecated)
+	case types.DepositStatus_Burned:
+		prefix = key.FromStr(burnedDepositPrefixDeprecated)
 	}
 
-	// migrate in batches so memory pressure doesn't become too large
-	keysToDelete := make([][]byte, 0, 1000)
-	for {
-		iterBurnerAddr := ck.getStore(ctx).Iterator(utils.KeyFromStr(oldKey))
+	iter := ck.getStore(ctx).IteratorNew(prefix)
+	defer utils.CloseLogError(iter, ck.Logger(ctx))
 
-		if !iterBurnerAddr.Valid() {
-			return iterBurnerAddr.Close()
+	for ; iter.Valid(); iter.Next() {
+		iteratedDepositCount++
+
+		var deposit types.ERC20Deposit
+		iter.UnmarshalValue(&deposit)
+
+		transferEvents := getTransferEventsByTxIDAndAddress(ctx, ck, deposit.TxID, deposit.BurnerAddress)
+		if len(transferEvents) == 0 {
+			// Deposits from the time when we had not started doing event processing.
+			// Their log indexes cannot be retrieved anymore and are therefore ignored.
+			ignoredDepositCount++
+			continue
 		}
-		var burnerInfo types.BurnerInfo
-		for ; iterBurnerAddr.Valid() && len(keysToDelete) < 1000; iterBurnerAddr.Next() {
-			iterBurnerAddr.UnmarshalValue(&burnerInfo)
-			ck.SetBurnerInfo(ctx, burnerInfo)
-			keysToDelete = append(keysToDelete, iterBurnerAddr.Key())
+
+		rawKey := iter.Key()
+		defer ck.getStore(ctx).DeleteRaw(rawKey)
+
+		for _, event := range transferEvents {
+			_, existingDepositStatus, ok := ck.GetDeposit(ctx, event.TxID, event.Index)
+			if ok && existingDepositStatus == status {
+				continue
+			}
+
+			if existingDepositStatus != status {
+				ck.Logger(ctx).Debug(fmt.Sprintf("deposit status changes from %s to %s", existingDepositStatus.String(), status.String()),
+					"chain", ck.GetName(),
+					"tx_id", event.TxID.Hex(),
+					"log_index", event.Index,
+				)
+			}
+
+			newDeposit := types.ERC20Deposit{
+				TxID:             event.TxID,
+				LogIndex:         event.Index,
+				Amount:           event.GetTransfer().Amount,
+				Asset:            deposit.Asset,
+				DestinationChain: deposit.DestinationChain,
+				BurnerAddress:    deposit.BurnerAddress,
+			}
+			ck.DeleteDeposit(ctx, newDeposit)
```

Comment:
we shouldn't delete while iterating

Reply:
atm anything is fine, but that is coupling our code to the concrete interface implementation. Since there is no guarantee that there won't be any overlap we should generally avoid deleting while iterating

Reply:
I can get it out of the iteration, but it feels silly cuz we are also doing `SetDeposit` in the loop and it does an equal amount of "damage" to the iterator if any. Unless you want to move both of them out?